src/modules/microdds_client/microdds_client.cpp

```diff
@@ -46,27 +45,66 @@
 #include <stdlib.h>
 #include <unistd.h>
 
+#if defined(CONFIG_NET) || defined(__PX4_POSIX)
+# define MICRODDS_CLIENT_UDP 1
+#endif
+
 #define STREAM_HISTORY  4
 #define BUFFER_SIZE (UXR_CONFIG_SERIAL_TRANSPORT_MTU * STREAM_HISTORY) // MTU==512 by default
 
 using namespace time_literals;
 
+void on_time(uxrSession *session, int64_t current_time, int64_t received_timestamp, int64_t transmit_timestamp,
+	     int64_t originate_timestamp, void *args)
+{
+	// latest round trip time (RTT)
+	int64_t rtt = current_time - originate_timestamp;
+
+	// HRT to AGENT
+	int64_t offset_1 = (received_timestamp - originate_timestamp) - (rtt / 2);
+	int64_t offset_2 = (transmit_timestamp - current_time) - (rtt / 2);
+
+	session->time_offset = (offset_1 + offset_2) / 2;
+
+	if (args) {
+		Timesync *timesync = static_cast<Timesync *>(args);
+		timesync->update(current_time / 1000, transmit_timestamp, originate_timestamp);
+
+		//fprintf(stderr, "time_offset: %ld, timesync: %ld, diff: %ld\n", session->time_offset/1000, timesync->offset(), session->time_offset/1000 + timesync->offset());
+
+		session->time_offset = -timesync->offset() * 1000; // us -> ns
+	}
+}
+
 MicroddsClient::MicroddsClient(Transport transport, const char *device, int baudrate, const char *host,
-			       const char *port, bool localhost_only)
-	: _localhost_only(localhost_only)
+			       const char *port, bool localhost_only, const char *client_namespace)
+	: _localhost_only(localhost_only), _client_namespace(client_namespace)
 {
 	if (transport == Transport::Serial) {
 
-		int fd = open(device, O_RDWR | O_NOCTTY | O_NONBLOCK);
+		int fd = -1;
 
-		if (fd < 0) {
-			PX4_ERR("open %s failed (%i)", device, errno);
+		for (int attempt = 0; attempt < 3; attempt++) {
+			fd = open(device, O_RDWR | O_NOCTTY | O_NONBLOCK);
+
+			if (fd < 0) {
+				PX4_ERR("open %s failed (%i)", device, errno);
+				// sleep before trying again
+				usleep(1'000'000);
+
+			} else {
+				break;
+			}
 		}
 
 		_transport_serial = new uxrSerialTransport();
 
 		if (fd >= 0 && setBaudrate(fd, baudrate) == 0 && _transport_serial) {
```

Comment:
I know this PR didn't introduce the setBaudrate code, but upon examination, it looks the implementation could be incompatible with existing configurations with baudrates higher than B921600 (B1000000 and higher). The RTPS implementation had support for 1M-4M, so maybe worth adding those base in here in this pass?

Reply:
Interesting, I didn't realize is was that high. Let's get it updated.